klippy/extras/axis_twist_compensation.py

```diff
@@ -0,0 +1,347 @@
+# Axis Twist Compensation
+
+# Copyright (C) 2022  Jeremy Tan <jeremytkw98@gmail.com>
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+"""
+[axis_twist_compensation]
+horizontal_move_z: 5
+speed: 50
+start_x: 0 ; nozzle's x coordinate at the start of the calibration ! required
+end_x: 200 ; nozzle's x coordinate at the end of the calibration ! required
+y: 100 ; nozzle's y coordinate during the calibration ! required
+type: multilinear
+"""
+
+import logging
+import math
+from . import manual_probe as ManualProbe, bed_mesh as BedMesh
+
+
+DEFAULT_N_POINTS = 3
+
+
+class Config:
+    DEFAULT_SPEED = 50.
+    DEFAULT_HORIZONTAL_MOVE_Z = 5.
+    REQUIRED = True
+    OPTIONAL = False
+    CONFIG_OPTIONS = {
+        'horizontal_move_z': (float, OPTIONAL, DEFAULT_HORIZONTAL_MOVE_Z),
+        'speed': (float, OPTIONAL, DEFAULT_SPEED),
+        'start_x': (float, REQUIRED, None),
+        'end_x': (float, REQUIRED, None),
+        'y': (float, REQUIRED, None),
+        'type': (str, OPTIONAL, 'multilinear'),
+        'z_compensations': (str, OPTIONAL, None),
+    }
+
+    @staticmethod
+    def save_to_config(name, z_compensations, configfile):
+        values_as_str = ', '.join([Helpers.format_float_to_n_decimals(x)
+                                   for x in z_compensations])
+        configfile.set(name, 'z_compensations', values_as_str)
+
+
+class AxisTwistCompensation:
+    def __init__(self, config):
+        # get printer
+        self.config = config
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+
+        # get values from [axis_twist_compensation] section in printer .cfg
+        for config_key, \
+            (config_type, required, default) in Config.CONFIG_OPTIONS.items():
+            value = None
+            if config_type == float:
+                value = config.getfloat(config_key, default)
+            elif config_key == 'z_compensations':
+                value = Helpers.parse_comma_separated_floats(
+                    config.get('z_compensations', default=""))
+            else:
+                value = config.get(config_key, default)
+            if required and value is None:
+                raise config.error(
+                    "Missing required config option for section [{}]: {}"
+                    .format(config.get_name(), config_key))
+            setattr(self, config_key, value)
```

Comment:
I find this code to be very confusing.  I think it would be much more readable if `config_getXXX()` was directly called the 7 times.

Please use `config.getlist()` (or `config.getlists()`) instead of implementing a custom `parse_comma_separated_floats()` function.

Don't keep a reference to the `config` object in the local class (doing so leads to confusion as all options must be read during the initialization phase).  If you need the `config.get_name()` value then read it during init and store it in a `self.name` member variable.

For what it is worth, I think the static `Config` and `Helpers` classes make the code harder to read.

Reply:
All fixed, but I have one question left. I now use config.getlist instead of parsing the line, but I couldn't find an equivalent to save the list. In bed_mesh.py the formatting is done by hand too: https://github.com/Klipper3d/klipper/blob/a96608add40e316f25f15d9c9d1c1fbd86dbbebe/klippy/extras/bed_mesh.py#L1239-L1245 

Is there a method to do that or should I leave the manual formatting code here?

Reply:
To save an updated list, just format it manually (eg, `",".join(["%.3f" % (v,) for v in my_values])`).

-Kevin